From 1ab38f4f7840a3c19bf961a24630a992a8373a76 Mon Sep 17 00:00:00 2001
From: Jon Dufresne <jon.dufresne@gmail.com>
Date: Wed, 14 Aug 2019 04:59:59 -0700
Subject: [PATCH] Simplify subprocess use with high level run API

Per the Python docs, subprocess.run() is the preferred interface for
running subprocess and replaces the older API (such as
subprocess.check_output())

https://docs.python.org/3/library/subprocess.html

> The recommended approach to invoking subprocesses is to use the run()
> function for all use cases it can handle. For more advanced use cases,
> the underlying Popen interface can be used directly.
>
>
> The run() function was added in Python 3.5; if you need to retain
> compatibility with older versions, see the Older high-level API
> section.

Allows for:

- Return a str from get_output() to decode once instead of every line of
  the result.

- Pass command arguments as a list of strings to ensure no possibility
  for unescaped shell injection.
---
 isort/hooks.py | 17 +++++++++--------
 test_isort.py  | 28 ++++++++++++++++++++--------
 2 files changed, 29 insertions(+), 16 deletions(-)

diff --git a/isort/hooks.py b/isort/hooks.py
index a14493f04..2e0752704 100644
--- a/isort/hooks.py
+++ b/isort/hooks.py
@@ -28,17 +28,18 @@
 from isort import SortImports
 
 
-def get_output(command: str) -> bytes:
+def get_output(command: List[str]) -> str:
     """
     Run a command and return raw output
 
     :param str command: the command to run
     :returns: the stdout output of the command
     """
-    return subprocess.check_output(command.split())
+    result = subprocess.run(command, stdout=subprocess.PIPE, check=True)
+    return result.stdout.decode()
 
 
-def get_lines(command: str) -> List[str]:
+def get_lines(command: List[str]) -> List[str]:
     """
     Run a command and return lines of output
 
@@ -46,7 +47,7 @@ def get_lines(command: str) -> List[str]:
     :returns: list of whitespace-stripped lines output by command
     """
     stdout = get_output(command)
-    return [line.strip().decode() for line in stdout.splitlines()]
+    return [line.strip() for line in stdout.splitlines()]
 
 
 def git_hook(strict=False, modify=False):
@@ -64,19 +65,19 @@ def git_hook(strict=False, modify=False):
     """
 
     # Get list of files modified and staged
-    diff_cmd = "git diff-index --cached --name-only --diff-filter=ACMRTUXB HEAD"
+    diff_cmd = ["git", "diff-index", "--cached", "--name-only", "--diff-filter=ACMRTUXB HEAD"]
     files_modified = get_lines(diff_cmd)
 
     errors = 0
     for filename in files_modified:
         if filename.endswith('.py'):
             # Get the staged contents of the file
-            staged_cmd = "git show :%s" % filename
+            staged_cmd = ["git", "show", ":%s" % filename]
             staged_contents = get_output(staged_cmd)
 
             sort = SortImports(
                 file_path=filename,
-                file_contents=staged_contents.decode(),
+                file_contents=staged_contents,
                 check=True
             )
 
@@ -85,7 +86,7 @@ def git_hook(strict=False, modify=False):
                 if modify:
                     SortImports(
                         file_path=filename,
-                        file_contents=staged_contents.decode(),
+                        file_contents=staged_contents,
                         check=False,
                     )
 
diff --git a/test_isort.py b/test_isort.py
index bb5305ac8..ff061da0f 100644
--- a/test_isort.py
+++ b/test_isort.py
@@ -24,9 +24,9 @@
 import os
 import os.path
 import posixpath
+import subprocess
 import sys
 import sysconfig
-from subprocess import check_output
 from tempfile import NamedTemporaryFile
 from typing import Any, Dict, List
 
@@ -3230,11 +3230,15 @@ def test_settings_path_skip_issue_909(tmpdir):
     test_run_directory = os.getcwd()
     os.chdir(str(base_dir))
     with pytest.raises(Exception):  # without the settings path provided: the command should not skip & identify errors
-        check_output(['isort', '--check-only'])
-    results = check_output(['isort', '--check-only', '--settings-path=conf/.isort.cfg'])
+        subprocess.run(['isort', '--check-only'], check=True)
+    result = subprocess.run(
+        ['isort', '--check-only', '--settings-path=conf/.isort.cfg'],
+        stdout=subprocess.PIPE,
+        check=True
+    )
     os.chdir(str(test_run_directory))
 
-    assert b'skipped 2' in results.lower()
+    assert b'skipped 2' in result.stdout.lower()
 
 
 def test_skip_paths_issue_938(tmpdir):
@@ -3261,16 +3265,24 @@ def test_skip_paths_issue_938(tmpdir):
 
     test_run_directory = os.getcwd()
     os.chdir(str(base_dir))
-    results = check_output(['isort', 'dont_skip.py', 'migrations/file_glob_skip.py'])
+    result = subprocess.run(
+        ['isort', 'dont_skip.py', 'migrations/file_glob_skip.py'],
+        stdout=subprocess.PIPE,
+        check=True,
+    )
     os.chdir(str(test_run_directory))
 
-    assert b'skipped' not in results.lower()
+    assert b'skipped' not in result.stdout.lower()
 
     os.chdir(str(base_dir))
-    results = check_output(['isort', '--filter-files', '--settings-path=conf/.isort.cfg', 'dont_skip.py', 'migrations/file_glob_skip.py'])
+    result = subprocess.run(
+        ['isort', '--filter-files', '--settings-path=conf/.isort.cfg', 'dont_skip.py', 'migrations/file_glob_skip.py'],
+        stdout=subprocess.PIPE,
+        check=True,
+    )
     os.chdir(str(test_run_directory))
 
-    assert b'skipped 1' in results.lower()
+    assert b'skipped 1' in result.stdout.lower()
 
 
 def test_failing_file_check_916():
